File: src/api/chatMappings.ts
/**
 * Kemory dashboard — Chat namespace mappings API (chats-v1 UI).
 *
 * Mappings let the user collapse multiple (platform, source project)
 * combinations onto one Kemory namespace. Many-to-one is by design
 * (multiple rows with the same `target_namespace`).
 *
 * Backend reference: backend/api/routes/chat_mappings.py
 *                    backend/services/ai_chat_service.py
 *                    (ChatMappingCreate / Update / Response)
 */
import { api } from './client'
import type { Platform } from './chats'

export interface ChatMappingResponse {
  mapping_id: string
  platform: Platform
  source_project_id: string | null
  source_project_name_pattern: string | null
  target_namespace: string
  priority: number
  enabled: boolean
  created_at: string
  updated_at: string
}

export interface ChatMappingCreateRequest {
  platform: Platform
  source_project_id?: string | null
  source_project_name_pattern?: string | null
  target_namespace: string
  priority?: number
  enabled?: boolean
}

export interface ChatMappingUpdateRequest {
  target_namespace?: string
  priority?: number
  enabled?: boolean
  source_project_id?: string | null
  source_project_name_pattern?: string | null
}

export async function listMappings(): Promise<ChatMappingResponse[]> {
  return api.get('api/v1/chat-mappings').json()
}

export async function createMapping(
  data: ChatMappingCreateRequest,
): Promise<ChatMappingResponse> {
  return api.post('api/v1/chat-mappings', { json: data }).json()
}

export async function updateMapping(
  mappingId: string,
  data: ChatMappingUpdateRequest,
): Promise<ChatMappingResponse> {
  return api.patch(`api/v1/chat-mappings/${encodeURIComponent(mappingId)}`, { json: data }).json()
}

export async function deleteMapping(mappingId: string): Promise<void> {
  await api.delete(`api/v1/chat-mappings/${encodeURIComponent(mappingId)}`)
}

File: backend/api/routes/chat_mappings.py
"""
Kemory — Chat namespace mapping CRUD (chats-v1).

Manages explicit ``(platform, project) → namespace`` overrides that the
AI Chats upsert path honours before falling through to the namespace
matcher. Many source projects can map to one namespace by inserting
multiple rows with the same ``target_namespace``.

Endpoints (all under ``/api/v1``):
  * POST   /chat-mappings              — create
  * GET    /chat-mappings              — list (own)
  * PATCH  /chat-mappings/{mapping_id} — update
  * DELETE /chat-mappings/{mapping_id} — delete
"""

from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from backend.core.auth import AuthContext, require_auth
from backend.core.database import get_db
from backend.services.ai_chat_service import (
    ChatMappingCreate,
    ChatMappingResponse,
    ChatMappingUpdate,
    create_mapping,
    delete_mapping,
    list_mappings,
    update_mapping,
)

router = APIRouter(prefix="/api/v1", tags=["AI Chats — Mappings"])


@router.post(
    "/chat-mappings",
    response_model=ChatMappingResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Create a chat namespace mapping override",
)
async def create_mapping_endpoint(
    request: ChatMappingCreate,
    auth: AuthContext = Depends(require_auth),
    db: AsyncSession = Depends(get_db),
):
    try:
        return await create_mapping(auth.user_id, auth.org_id, request, db)
    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(exc),
        ) from exc


@router.get(
    "/chat-mappings",
    response_model=list[ChatMappingResponse],
    summary="List own chat namespace mappings",
)
async def list_mappings_endpoint(
    auth: AuthContext = Depends(require_auth),
    db: AsyncSession = Depends(get_db),
):
    return await list_mappings(auth.user_id, db)


@router.patch(
    "/chat-mappings/{mapping_id}",
    response_model=ChatMappingResponse,
    summary="Update a chat namespace mapping",
)
async def update_mapping_endpoint(
    mapping_id: uuid.UUID,
    request: ChatMappingUpdate,
    auth: AuthContext = Depends(require_auth),
    db: AsyncSession = Depends(get_db),
):
    try:
        return await update_mapping(mapping_id, auth.user_id, request, db)
    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        ) from exc


@router.delete(
    "/chat-mappings/{mapping_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete a chat namespace mapping",
)
async def delete_mapping_endpoint(
    mapping_id: uuid.UUID,
    auth: AuthContext = Depends(require_auth),
    db: AsyncSession = Depends(get_db),
):
    try:
        await delete_mapping(mapping_id, auth.user_id, db)
    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        ) from exc
